[v24.2.x] Fix stepping down on timeout

mmaslankaprv opened the pull request:
Backport of PR https://github.com/redpanda-data/redpanda/pull/24590 
Fixes: #24668

mmaslankaprv committed:
r/tests: made election timeout and hb interval configurable

Made leader election timeout and heartbeat interval properties
runtime configurable in Raft fixture tests.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

bashtanov committed:
tests/gtest_raft_rpunit: monitor_test_fixture honour leadership changes

1) added a utility function to raft_fixture to execute a testing coro in
retry_with_leader
2) made both monitor_test_fixture tests use it

(cherry picked from commit d1ecf0e4debae17fb500a8708f38ae5c0314baa5)

mmaslankaprv committed:
raft: renamed reply_result::timeout to follower_busy

The `raft::reply_result::follower_busy` is indicating that the follower
was unable to process the heartbeat fast enough to generate a response.
Renaming the reply from `timeout` will make it less confusing for the
reader and differentiate the error code from an RPC timeout.

Signed-off-by: Michał Maślanka <michal@redpanda.com>
(cherry picked from commit 6a1e34beaddd5182ad2ff58beadc014480c0ea13)

mmaslankaprv committed:
storage/ntp_config: added copy method to ntp_config

Signed-off-by: Michał Maślanka <michal@redpanda.com>
(cherry picked from commit 95a29dba65fa96ad67cbf84de601e44bec970f8a)

mmaslankaprv committed:
raft/tests: wired in rpc service into the raft_fixture

Wired raft RPC service handler into Raft fixture to make the tests more
accurate and cover the service code with tests.

Signed-off-by: Michał Maślanka <michal@redpanda.com>
(cherry picked from commit 5f69d9b733a8c4c42282bebaa2667e8d0baaaf65)

mmaslankaprv committed:
r/tests: handle timeout in in memory rpc protocol

Propagating timeout to the node sending RPC request is crucial for
accurate testing of Raft implementation.

Signed-off-by: Michał Maślanka <michal@redpanda.com>
(cherry picked from commit 7d33bb5659e05a8bd480c4cc25d20a5cd7907f7c)

mmaslankaprv committed:
r/tets: added failure injectable log

Added a wrapper around the `storage::log` allowing us to inject storage
layer failures in Raft fixture tests.

Signed-off-by: Michał Maślanka <michal@redpanda.com>
(cherry picked from commit f04995a75105fb5e070accdc28e3b3de4e914836)

mmaslankaprv committed:
r/heartbeat_manager: do not treat follower busy err as missed heartbeat

When follower is busy it may fail fast processing full heartbeat
requests sent by the leader. In this case a follower RPC handler sets
the `follower_busy` result in heartbeat_reply. Leader should still treat
a follower replica as online in this case. The replica hosting node must
be online to reply with the `follower_busy` error.

This way we prevent to eager leader step downs when follower replicas
are slow.
Signed-off-by: Michał Maślanka <michal@redpanda.com>
(cherry picked from commit 8b57b4210166d2e8804c3da39421f79951228769)

mmaslankaprv committed:
r/tests: added test checking if leadership is stable after timeout

Signed-off-by: Michał Maślanka <michal@redpanda.com>
(cherry picked from commit 67e7c6ea213704f8c3edb30028a9f3a0da517da3)

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/60349-->
<!--27ba5862-2490-4433-afe1-ac5a6fb68172-->
<!--title-->
## Retry command for [Build#60349](https://buildkite.com/redpanda/redpanda/builds/60349)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"enable_failures":false,"mixed_versions":true,"with_tiered_storage":true}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--7089277b-6f01-4f3d-92e1-e582c6f7b92d-->
<!--TEST RESULTS-->
<!--d99e8510-f2fd-4c01-8d95-7405d025a66b-->
<!--TEST RESULTS-->
<!--5b2e2250-20a4-4934-82b1-fec062a29941-->
<!--TEST RESULTS-->
<!--45d8ef32-a2ff-4361-9912-1613d0e5fd00-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60349">build#60349</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:----------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| distributed_kv_stm_tests_rpunit.distributed_kv_stm_tests_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/60349#01944155-5086-44b3-92bc-601f600f4fe7 | FAIL | 0/2 |
| distributed_kv_stm_tests_rpunit.distributed_kv_stm_tests_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/60349#01944155-5086-4c19-949e-47ae1823c063 | FAIL | 0/2 |
| gtest_archival_rpunit.gtest_archival_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/60349#01944155-5086-44b3-92bc-601f600f4fe7 | FAIL | 0/2 |
| gtest_archival_rpunit.gtest_archival_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/60349#01944155-5086-4c19-949e-47ae1823c063 | FAIL | 0/2 |
| gtest_raft_rpunit.gtest_raft_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/60349#01944155-5086-44b3-92bc-601f600f4fe7 | FAIL | 0/2 |
| gtest_raft_rpunit.gtest_raft_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/60349#01944155-5086-4c19-949e-47ae1823c063 | FAIL | 0/2 |
| id_allocator_stm_test_rpunit.id_allocator_stm_test_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/60349#01944155-5086-44b3-92bc-601f600f4fe7 | FAIL | 0/2 |
| id_allocator_stm_test_rpunit.id_allocator_stm_test_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/60349#01944155-5086-4c19-949e-47ae1823c063 | FAIL | 0/2 |
| rptest.tests.delete_records_test.DeleteRecordsTest.test_delete_records_concurrent_truncations.cloud_storage_enabled=True.truncate_point=start_offset | ducktape | https://buildkite.com/redpanda/redpanda/builds/60349#0194419d-b830-4b8b-a662-0ebc1e076cbd | FLAKY | 5/6 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=False.mixed_versions=True.with_tiered_storage=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/60349#01944199-9127-4a22-8ab6-fefca6176271 | FAIL | 0/1 |
| tm_stm_tests_rpunit.tm_stm_tests_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/60349#01944155-5086-44b3-92bc-601f600f4fe7 | FAIL | 0/2 |
| tm_stm_tests_rpunit.tm_stm_tests_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/60349#01944155-5086-4c19-949e-47ae1823c063 | FAIL | 0/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60501">build#60501</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| gtest_raft_rpunit.gtest_raft_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/60501#01944b32-7230-45de-8a59-5a2c3b7b0a62 | FAIL | 0/2 |
| gtest_raft_rpunit.gtest_raft_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/60501#01944b32-7231-46dc-af2d-6bd3fa1ad9ba | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60632">build#60632</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| gtest_raft_rpunit.gtest_raft_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/60632#01945eba-3e44-4a67-98cb-fb56c31c5177 | FAIL | 0/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60948">build#60948</a></summary>


| test_id                             | test_kind   | job_url                                                                                   | test_status   | passed   |
|:------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| gtest_raft_rpunit.gtest_raft_rpunit | unit        | https://buildkite.com/redpanda/redpanda/builds/60948#019482cf-bdfb-4d76-9afd-edcf0b537b86 | FLAKY         | 1/2      |
</details>
<!--/body-->

mmaslankaprv committed:
r/r_monitor_test: fixed truncation on leader change test

Fixed previously unstable test. Now the test simply blocks append entry
requests from leader instead of relying on uncertain timeouts. Added
waiting for enqueue of replicate requests to make sure the requests
landed in the buffer before the leadership changed.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

ztlpn commented on src/v/storage/ntp_config.h:
```diff
@@ -299,6 +299,15 @@ class ntp_config {
+          _base_dir,
+          _overrides ? std::make_unique<default_overrides>(*_overrides)
+                     : nullptr,
+          _revision_id};
```
`_initial_rev` as well?

ztlpn commented on src/v/raft/tests/raft_fixture.h:
```diff
@@ -249,6 +269,12 @@ class raft_node_instance : public ss::weakly_referencable<raft_node_instance> {
+    service_t& get_service() { return _service; }
+
+
+    service_t& get_service() { return _service; }
```
duplicate methods

ztlpn commented on src/v/raft/tests/raft_fixture.cc:
```diff
@@ -401,13 +413,24 @@ raft_node_instance::raft_node_instance(
+  , _election_timeout(config::mock_binding<std::chrono::milliseconds>(
+      std::move(election_timeout)))
+  , _heartbeat_interval(config::mock_binding<std::chrono::milliseconds>(
+      std::move(heartbeat_interval)))
```
hmm where is this parameter coming from?

ztlpn commented on src/v/raft/tests/replication_monitor_tests.cc:
```diff
@@ -49,75 +43,112 @@ auto populate_waiters(
+            if (id == leader.get_vnode().id()) {
+                node->on_dispatch([](model::node_id, raft::msg_type mt) {
+                    if (mt == raft::msg_type::append_entries) {
+                        throw std::runtime_error("dropping_append_entries");
```
hmm should this be committed to dev as well?
